### PR TITLE
2 Point calibration fix

### PR DESCRIPTION
**Pull request by redoxcode**
This PR should fix an oversight in the way the calibration values are applied when using only 2 Points.
The issue was described here: https://github.com/Allen-Synthesis/EuroPi/issues/221#issuecomment-1446950213
In short: If we are not mistaken the adc offset measured when 0V is applied, is currently not taken into account.
Old:
result = 'adc raw value' / 'calibration at 10V'
New:
result = ('adc raw value' - 'calibration at 0V') / ('calibration at 10V' - 'calibration at 0V)

(some tests still fail, but it's only one contrib script (tests/contrib/test_europi_turing_machine.py) so I guess it's an issue of this test and not the code changes)

### CODE REVIEW

**Comment by awonak** [comment]
@redoxcode do you have some time to sync this branch to HEAD, and fix the lint and test errors? We'd love to include this in our next release.

**Comment by redoxcode** [comment]
@awonak sync is done. I will try to work out why tests fail

**Comment by redoxcode** [comment]
ok, there where some bugs in the implementation and some corner cases I missed (like when the input is below the 0 point calibration). But I think its good now!
But please check on actual hardware and do the calibration.
I also had to fix the 2 point calibration in the mock hardware.
And some how my black output was different than what lint wants :thinking:

File: software/firmware/europi.py [hunk 1]
```diff
@@ -213,16 +213,26 @@ def percent(self, samples=None):
         """Current voltage as a relative percentage of the component's range."""
         # Determine the percent value from the max calibration value.
         reading = self._sample_adc(samples)
-        max_value = max(reading, INPUT_CALIBRATION_VALUES[-1])
-        return reading / max_value
+        max_value = max(
+            reading - INPUT_CALIBRATION_VALUES[0],
+            INPUT_CALIBRATION_VALUES[-1] - INPUT_CALIBRATION_VALUES[0],
+        )
+        return max((reading - INPUT_CALIBRATION_VALUES[0]) / max_value, 0.0)
```
Comment (roryjamesallen) [inline software/firmware/europi.py:220]: the calculation of ```reading - INPUT_CALIBRATION_VALUES[0]``` is made twice, I'd suggest making the calculation between lines 215 and 216 with something like
```offset = reading - INPUT_CALIBRATION_VALUES[0]``` and then refer to that offset

File: software/firmware/europi.py [hunk 2]
```diff
@@ -213,16 +213,26 @@ def percent(self, samples=None):
...
 
     def read_voltage(self, samples=None):
         reading = self._sample_adc(samples)
-        max_value = max(reading, INPUT_CALIBRATION_VALUES[-1])
-        percent = reading / max_value
+        max_value = max(
+            reading - INPUT_CALIBRATION_VALUES[0],
+            INPUT_CALIBRATION_VALUES[-1] - INPUT_CALIBRATION_VALUES[0],
+        )
+        percent = max((reading - INPUT_CALIBRATION_VALUES[0]) / max_value, 0.0)
```
Comment (roryjamesallen) [inline software/firmware/europi.py:228]: The same duplicated code comment as above

**Review by roryjamesallen — CHANGES_REQUESTED** [review]

**Comment by awonak** [comment]
LGTM, thank you for improving the calibration process @redoxcode & @francoisgeorgy
